lightning/src/ln/channelmanager.rs

```diff
@@ -7500,51 +7541,60 @@ where
+						match htlc_source {
+							HTLCSource::PreviousHopData(prev_hop_data) => {
+								let pending_forward_matches_htlc = |info: &PendingAddHTLCInfo| {
+									info.prev_funding_outpoint == prev_hop_data.outpoint &&
+										info.prev_htlc_id == prev_hop_data.htlc_id
+								};
+								// The ChannelMonitor is now responsible for this HTLC's
+								// failure/success and will let us know what its outcome is. If we
+								// still have an entry for this HTLC in `forward_htlcs` or
+								// `pending_intercepted_htlcs`, we were apparently not persisted after
+								// the monitor was when forwarding the payment.
+								forward_htlcs.retain(|_, forwards| {
+									forwards.retain(|forward| {
+										if let HTLCForwardInfo::AddHTLC(htlc_info) = forward {
+											if pending_forward_matches_htlc(&htlc_info) {
+												log_info!(args.logger, "Removing pending to-forward HTLC with hash {} as it was forwarded to the closed channel {}",
+													log_bytes!(htlc.payment_hash.0), log_bytes!(monitor.get_funding_txo().0.to_channel_id()));
+												false
+											} else { true }
 										} else { true }
+									});
+									!forwards.is_empty()
+								});
+								pending_intercepted_htlcs.as_mut().unwrap().retain(|intercepted_id, htlc_info| {
+									if pending_forward_matches_htlc(&htlc_info) {
+										log_info!(args.logger, "Removing pending intercepted HTLC with hash {} as it was forwarded to the closed channel {}",
+											log_bytes!(htlc.payment_hash.0), log_bytes!(monitor.get_funding_txo().0.to_channel_id()));
+										pending_events_read.retain(|event| {
+											if let Event::HTLCIntercepted { intercept_id: ev_id, .. } = event {
+												intercepted_id != ev_id
+											} else { true }
+										});
+										false
 									} else { true }
 								});
-								!forwards.is_empty()
-							});
-							pending_intercepted_htlcs.as_mut().unwrap().retain(|intercepted_id, htlc_info| {
-								if pending_forward_matches_htlc(&htlc_info) {
-									log_info!(args.logger, "Removing pending intercepted HTLC with hash {} as it was forwarded to the closed channel {}",
-										log_bytes!(htlc.payment_hash.0), log_bytes!(monitor.get_funding_txo().0.to_channel_id()));
-									pending_events_read.retain(|event| {
-										if let Event::HTLCIntercepted { intercept_id: ev_id, .. } = event {
-											intercepted_id != ev_id
-										} else { true }
-									});
-									false
-								} else { true }
-							});
+							},
+							HTLCSource::OutboundRoute { payment_id, session_priv, path, .. } => {
+								if let Some(preimage) = preimage_opt {
+									let pending_events = Mutex::new(pending_events_read);
+									// Note that we set `from_onchain` to "false" here,
+									// deliberately keeping the pending payment around forever.
+									// Given it should only occur when we have a channel we're
+									// force-closing for being stale that's okay.
```

Comment:
Are we doing this because it's more accurate, since the payment was not claimed on-chain? And does this edge case cause us to swallow a `PaymentPathSuccessful` event that would otherwise be issued (perhaps this is fixed in followup)?